### PR TITLE
Renamed Unrestricted to Ur

### LINKED ISSUE #131: The name of Unrestricted
This is something which has bothered everyone for months. It resurfaced recently in our internal discussion channels. We still don't have a clear solution.

The problem: `Unrestricted`, while a perfectly descriptive name, is very long. And considering how often this type is referred to in type signatures (and its constructor in function definition), we really want a shorter name. It's for everybody's sanity.

This naming issue can also be accompanied  by a reflection of how to name of the accompanying, and commonly useful

```haskell
data X s a where
  X :: s #-> a -> X s a
```
(the fields may be in the reverse order, I don't care much either way)

Here is a list of all the proposals that have been mooted that I can remember

- `Many`: but since it's already used for the name of the multiplicity, it invites many conflicts which are likely to be unpleasant.
- `Bang` after the fact that `Unrestricted a` is write `!a` in traditional linear logic literature. But it's fairly meaningless, if you don't know that notation.
- `Poly`: after _polynomial_ (linear: I must consume once, quadratic: I must consume twice, … polynomial: either of these). It does makes sense from a point of view of the mathematical analogy. And `Poly` means something like “several” in Greek, so it makes sense.
- `U`: first letter of `Unrestricted`. Easy to use, but a one-letter name may be consider a bit too rare and precious for this use.
- `This`: short, but ultimately hard to convey an intuition. It sounds a bit like it ought to mean the identity, which I don't think is a very good intuition. It could work if we have a story for how this makes sense. It does have the advantage that it comes with a reasonable name for `X`: ``s `AndThis` a``.
- `Lax`: mostly a synonymous of `Unrestricted`. But nobody's really convinced by it
- `Freely`, `Surely`: adverbs in the tradition of using such adverbs to pronounce modality (sometime `!a` is pronounced “certainly `a`”, see also “there merely exists” in homotopy type theory).
  - It could be a corresponding adjective, though `Free` is already taken, and `Sure` sounds weird. (like is “p is a mere proposition”)
  - Or an adverb of the same kind as `Maybe`
- `Box`: very generic
- `With`: this one has a bit of a twist, we would have 
   ```haskell
   type With :: Multiplicity -> Type -> Type
   data With p a where
     With :: a #p-> With p a
   ```
   So that `With 'Many a` is `Unrestricted a`. But since this last case is so common we would make a shorthand, such as `With'`.
- `More`: A bit like `Many`. Though it may sound a bit excessive, I guess.

I am, personally, fully convinced by none of these. But they are good starting points. Naming is hard. But, sometimes, after stumbling in the dark for a while, we suddenly fine the name which should have been obvious all along. Help me find it.

### PR DESCRIPTION
**Pull request by Divesh-Otwani**
This PR renames `Unrestricted` to `Ur` and closes #131. 

The discussion of how we arrived at this name is documented in the issue. Basically, `Unrestricted` is a really good name but too long. Since this data type is so fundamental to linear programming that any user will quickly become familiar with its definition, it's sensible to keep the name short for line length.

Remarks:
- In the spirit of keeping things short, I've changed qualified imports that were `Unrestricted` to `Ur`. Is this problematic or can we allow it since it will be so common? (I'm leaning towards the latter.)
- `unUnrestricted` has been renamed to `unUr`

### CODE REVIEW

File: src/Data/Unrestricted/Linear.hs
Comment (mboes) [inline src/Data/Unrestricted/Linear.hs:118]: `Ur` is itself a contraction. So I'd even go for `unur` here. Simple easy-to-spell and type combinator.

**Review by utdemir — APPROVED** [review]
LGTM (after addressing Mathieu's comment).

**Comment by aspiwack** [comment]
> * `unUnrestricted` has been renamed to `unUr`

Sounds like the right choice. `unur` as @mboes is just fine too.

> * In the spirit of keeping things short, I've changed qualified imports that were `Unrestricted` to `Ur`. Is this problematic or can we allow it since it will be so common? (I'm leaning towards the latter.)

I'm less inclined to this.But let's try it and see how it pans out in practice.

File: docs/USER_GUIDE.md
Comment (aspiwack) [inline docs/USER_GUIDE.md:92]: Ah, there is a bit of a mistake here. The proper pattern is `f :: X -> (SomeType #-> Ur b) #-> Ur b` (fix it after you merge, it's unrelated to this PR).

**Review by aspiwack — APPROVED** [review]

File: examples/Simple/Pure.hs
Comment (Divesh-Otwani) [reply examples/Simple/Pure.hs:303]: Thx :smile:

File: src/Data/Unrestricted/Linear.hs
Comment (Divesh-Otwani) [reply src/Data/Unrestricted/Linear.hs:23]: Yes, agreed.

File: src/Data/Unrestricted/Linear.hs
Comment (Divesh-Otwani) [reply src/Data/Unrestricted/Linear.hs:118]: Cool :sunglasses:

**Pull request merged**
